### PR TITLE
improve zero grad communication overlap with pp

### LINKED ISSUE #99: [Feature] only overlap sync_grad in pp0 with pipeline parallelism
### Describe the feature

image
only overlap sync_grad in pp0 with pipeline parallelism

if the network is poor, sync_grad may be the main performance bottleneck, and the pipeline bubble could be huge if we overlap sync_grad with computation, since pp0 must wait the communication of other stages in current implementation

### Will you implement it?

- [ ] I would like to implement this feature and create a PR!

### PR DESCRIPTION
**Pull request by mwiacx**
## improve zero grad communication overlap with pp

 When pipeline parallelism is enabled, we prefer to only enable optimizer gradient communication overlap in the first stage, to avoid amplifying the communication overhead stage by stage in cases where the optimizer communication overhead is greater than the compute overhead. For pipeline stages except the first, even if overlap is not enabled, their gradient synchronization overhead can be well hidden by the inherent bubbles of pipeline parallelism.

closes #99

### CODE REVIEW

File: internlm/train/pipeline.py [hunk 1]
```diff
@@ -304,19 +302,38 @@ def initialize_optimizer(model: Union[nn.Module, nn.ModuleList], isp_communicato
...
+        # gradient communication overlap in the first stage, to avoid amplifying
+        # the communication overhead stage by stage in cases where the optimizer
+        # communication overhead is greater than the compute overhead.
+        # For pipeline stages except the first, even if overlap is not enabled,
+        # their gradient synchronization overhead can be well hidden by
+        # the inherent bubbles of pipeline parallelism.
+        zero_cfg.overlap_sync_grad = False
+
+    if zero_cfg.overlap_sync_param:
+        param_bcast_sync_handler = ParamAsyncBcastHandler(ParallelMode.ZERO1, model, isp_communicator)
```
Comment (sunpengsdu) [inline internlm/train/pipeline.py:320]: but this is for parameter broadcast, not gradient synchronisation with allreduce?

**Review by sunpengsdu — APPROVED** [review]
